File: server/services/payment-service/src/utils/httpClient.ts
import axios from 'axios';
import dotenv from 'dotenv';

dotenv.config();

const ORDER_SERVICE_URL = (process.env.ORDER_SERVICE_URL || 'http://order-service:8000').replace(/\/$/, '');
const CART_SERVICE_URL = (process.env.CART_SERVICE_URL || 'http://cart-service:8000').replace(/\/$/, '');

export class MicroserviceClient {
  /**
   * Notify Order Service of payment status change (SUCCESS or FAILED)
   */
  static async notifyOrderService(orderId: string, paymentStatus: 'SUCCESS' | 'FAILED', transactionId?: string): Promise<boolean> {
    const url = `${ORDER_SERVICE_URL}/orders/${orderId}/payment-status`;
    try {
      const response = await axios.patch(
        url,
        {
          status: paymentStatus,
          transaction_id: transactionId
        },
        { timeout: 8000 }
      );
      console.log(`[Payment -> Order] Successfully updated order_id=${orderId} payment status to ${paymentStatus}`);
      return response.status === 200;
    } catch (error: any) {
      console.error(`[Payment -> Order] Failed to update order_id=${orderId}: ${error.response?.data?.detail || error.message}`);
      return false;
    }
  }

  /**
   * Notify Cart Service to clear user's purchased cart items
   */
  static async clearUserCart(userId: string): Promise<boolean> {
    const url = `${CART_SERVICE_URL}/api/carts/user/${userId}`;
    try {
      const response = await axios.delete(url, { timeout: 8000 });
      console.log(`[Payment -> Cart] Successfully cleared cart for user_id=${userId}`);
      return response.status === 200;
    } catch (error: any) {
      console.error(`[Payment -> Cart] Failed to clear cart for user_id=${userId}: ${error.response?.data?.detail || error.message}`);
      return false;
    }
  }
}

File: server/services/payment-service/app/core/http_client.py
import logging
import httpx
from app.core.config import ORDER_SERVICE_URL, CART_SERVICE_URL

logger = logging.getLogger(__name__)


class ServiceHTTPClient:
    @staticmethod
    async def notify_order_service(order_id: str, status: str, transaction_id: str = None) -> bool:
        url = f"{ORDER_SERVICE_URL}/orders/{order_id}/payment-status"
        payload = {
            "status": status,
            "transaction_id": transaction_id
        }
        try:
            async with httpx.AsyncClient(timeout=8.0) as client:
                response = await client.patch(url, json=payload)
            if response.status_code == 200:
                logger.info(f"[Payment -> Order] Updated order_id={order_id} status to {status}")
                return True
            else:
                logger.error(f"[Payment -> Order] Failed updating order_id={order_id}: {response.text}")
                return False
        except Exception as e:
            logger.error(f"[Payment -> Order] Exception notifying order_id={order_id}: {e}")
            return False

    @staticmethod
    async def clear_user_cart(user_id: str) -> bool:
        url = f"{CART_SERVICE_URL}/api/carts/user/{user_id}"
        try:
            async with httpx.AsyncClient(timeout=8.0) as client:
                response = await client.delete(url)
            if response.status_code == 200:
                logger.info(f"[Payment -> Cart] Cleared cart for user_id={user_id}")
                return True
            else:
                logger.error(f"[Payment -> Cart] Failed clearing cart for user_id={user_id}: {response.text}")
                return False
        except Exception as e:
            logger.error(f"[Payment -> Cart] Exception clearing cart for user_id={user_id}: {e}")
            return False
